src/engine/sync/abletonlink.cpp

```diff
@@ -0,0 +1,217 @@
+#include "engine/sync/abletonlink.h"
+
+#include <QtDebug>
+#include <cmath>
+
+#include "control/controlobject.h"
+#include "engine/sync/enginesync.h"
+#include "moc_abletonlink.cpp"
+#include "preferences/usersettings.h"
+#include "util/logger.h"
+
+namespace {
+const mixxx::Logger kLogger("AbletonLink");
+constexpr mixxx::Bpm kDefaultBpm(9999.9);
+} // namespace
+
+AbletonLink::AbletonLink(const QString& group, EngineSync* pEngineSync)
+        : m_group(group),
+          m_pEngineSync(pEngineSync),
+          m_syncMode(SyncMode::None),
+          m_oldTempo(kDefaultBpm),
+          m_audioBufferTimeMicros(0),
+          m_absTimeWhenPrevOutputBufferReachesDac(0),
+          m_sampleTimeAtStartCallback(0),
+          m_pLink(std::make_unique<ableton::BasicLink<MixxxClockRef>>(120.0)),
+          m_pLinkButton(std::make_unique<ControlPushButton>(ConfigKey(group, "sync_enabled"))),
+          m_pNumLinkPeers(std::make_unique<ControlObject>(ConfigKey(group, "num_peers"))) {
+    m_timeAtStartCallback = m_pLink->clock().micros();
+
+    m_pLinkButton->setButtonMode(mixxx::control::ButtonMode::Toggle);
+    m_pLinkButton->setStates(2);
+
+    connect(m_pLinkButton.get(),
+            &ControlObject::valueChanged,
+            this,
+            &AbletonLink::slotControlSyncEnabled,
+            Qt::DirectConnection);
+
+    m_pNumLinkPeers->setReadOnly();
+    m_pNumLinkPeers->forceSet(0);
+
+    // The callback is invoked on a Link - managed thread.
+    // The callback is the only entity, which access m_pNumLinkPeers
+    m_pLink->setNumPeersCallback([this](std::size_t numPeers) {
+        m_pNumLinkPeers->forceSet(numPeers);
+    });
```

Comment:
then make it a member of the lambda instead.
```suggestion
    // The callback is invoked on a Link - managed thread.
    m_pLink->setNumPeersCallback([ControlObject numLinkPeers = []{
        ControlObject numLinkPeers(ConfigKey(group, numPeers));
        numLinkPeers.setReadOnly();
        numLinkPeers.forceSet(0);
        return numLinkPeers;
    }()](std::size_t numPeers) {
        numLinkPeers.forceSet(numPeers);
    });
```

Reply:
The proposed code does not compile, as it does not match to the declaration in the headers of Link.

Reply:
that does not make sense. Can you post the error?

Reply:
![grafik](https://github.com/user-attachments/assets/421e4425-13aa-4c68-819c-7c47a3a41536)

Reply:
Thanks. looks like a syntax issue. I'll figure it out.

Reply:
init captures don't support explicit type declarations. Try this:
```suggestion
    // The callback is invoked on a Link - managed thread.
    m_pLink->setNumPeersCallback([numLinkPeers = []{
        ControlObject numLinkPeers(ConfigKey(group, numPeers));
        numLinkPeers.setReadOnly();
        numLinkPeers.forceSet(0);
        return numLinkPeers;
    }()](std::size_t numPeers) {
        numLinkPeers.forceSet(numPeers);
    });
```

Reply:
This doesn't work too.

Reply:
```suggestion
    // The callback is invoked on a Link - managed thread.
    m_pLink->setNumPeersCallback([numLinkPeers = [&]{
        ControlObject numLinkPeers(ConfigKey(group, numPeers));
        numLinkPeers.setReadOnly();
        numLinkPeers.forceSet(0);
        return numLinkPeers;
    }()](std::size_t numPeers) mutable {
        numLinkPeers.forceSet(numPeers);
    });
```

Reply:
Doesn't work:
```
  FAILED: CMakeFiles/mixxx-lib.dir/src/engine/sync/abletonlink.cpp.obj 
  C:\PROGRA~1\MICROS~1\2022\COMMUN~1\VC\Tools\MSVC\1441~1.341\bin\Hostx64\x64\cl.exe  /nologo /TP -DAMD64 -DDJINTEROP_SOURCE -DLINK_PLATFORM_WINDOWS=1 -DMIXXX_BUILD_RELEASE -DMIXXX_DEBUG_ASSERTIONS_ENABLED -DMIXXX_DEBUG_ASSERTIONS_FATAL -DMIXXX_USE_QML -DMIXXX_USE_QOPENGL -DNOMINMAX -DPROTOBUF_USE_DLLS -DQT_CONCURRENT_LIB -DQT_CORE5COMPAT_LIB -DQT_CORE_LIB -DQT_GUI_LIB -DQT_LABSQMLMODELS_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_OPENGL_LIB -DQT_PRINTSUPPORT_LIB -DQT_QMLINTEGRATION_LIB -DQT_QMLMODELS_LIB -DQT_QMLWORKERSCRIPT_LIB -DQT_QML_LIB -DQT_QUICKCONTROLS2IMPL_LIB -DQT_QUICKCONTROLS2_LIB -DQT_QUICKLAYOUTS_LIB -DQT_QUICKSHAPES_LIB -DQT_QUICKTEMPLATES2_LIB -DQT_QUICKWIDGETS_LIB -DQT_QUICK_LIB -DQT_SHADERTOOLS_LIB -DQT_SQL_LIB -DQT_SVGWIDGETS_LIB -DQT_SVG_LIB -DQT_TABLET_SUPPORT -DQT_TESTCASE_BUILDDIR=\"D:/mixxx/build/x64__portable\" -DQT_TESTCASE_SOURCEDIR=\"D:/mixxx\" -DQT_TESTLIB_LIB -DQT_USE_QSTRINGBUILDER -DQT_WIDGETS_LIB -DQT_XML_LIB -DSFC_SUPPORTS_SET_COMPRESSION_LEVEL -DUNICODE -DWIN32 -DWIN32_LEAN_AND_MEAN -DWIN64 -DWINVER=0x0601 -D_ATL_MIN_CRT -D_CRT_SECURE_NO_WARNINGS -D_ENABLE_EXTENDED_ALIGNED_STORAGE -D_SILENCE_CXX17_ITERATOR_BASE_CLASS_DEPRECATION_WARNING -D_UNICODE -D_USE_MATH_DEFINES -D_WIN32_WINNT=0x0601 -D_WIN64 -D__BATTERY__ -D__BROADCAST__ -D__BULK__ -D__ENGINEPRIME__ -D__FFMPEG__ -D__HID__ -D__HSS1394__ -D__LILV__ -D__MAD__ -D__MEDIAFOUNDATION__ -D__MODPLUG__ -D__OPUS__ -D__PORTMIDI__ -D__QTKEYCHAIN__ -D__RUBBERBAND__ -D__SNDFILE__ -D__SQLITE3__ -D__SSE2__ -D__SSE__ -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__STEM__ -D__VINYLCONTROL__ -D__WINDOWS__ -D__WV__ -ID:\mixxx\build\x64__portable\mixxx-lib_autogen\include -ID:\mixxx\src -ID:\mixxx\build\x64__portable\src -external:ID:\mixxx\lib\fidlib -external:ID:\mixxx\lib\portaudio -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include -external:ID:\mixxx\lib\rigtorp\SPSCQueue\include -external:ID:\mixxx\lib\replaygain -external:ID:\mixxx\lib\reverb -external:ID:\mixxx\lib\libshout-idjc\include -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\share\ableton-link\..\..\include\ableton -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\share\asio\..\..\include -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\djinterop -external:ID:\mixxx\lib\rekordbox-metadata -external:ID:\mixxx\lib\kaitai -external:ID:\mixxx\lib\mp3guessenc-0.27.4 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtConcurrent -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtCore -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\share\Qt6\mkspecs\win32-msvc -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtGui -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtNetwork -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtOpenGL -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtPrintSupport -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtWidgets -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQml -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlIntegration -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtSql -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtSvg -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtTest -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtXml -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtShaderTools -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtSvgWidgets -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtCore5Compat -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuick -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlModels -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtLabsQmlModels -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlModels\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlModels\6.5.3\QtQmlModels -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtCore\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtCore\6.5.3\QtCore -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQml\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQml\6.5.3\QtQml -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickControls2 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickControls2Impl -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickLayouts -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtGui\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtGui\6.5.3\QtGui -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuick\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuick\6.5.3\QtQuick -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickShapes -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickShapes\6.5.3 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickShapes\6.5.3\QtQuickShapes -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickTemplates2 -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQuickWidgets -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\Qt6\QtQmlWorkerScript -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\taglib -external:ID:\mixxx\lib\qm-dsp -external:ID:\mixxx\lib\qm-dsp\include -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\opus -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\hidapi -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\libusb-1.0 -external:ID:\mixxx\lib\xwax -external:ID:\mixxx\buildenv\mixxx-deps-2.5-x64-windows-06fe9b5\installed\x64-windows\include\wavpack -external:W0 /DWIN32 /D_WINDOWS /EHsc /Zc:__cplusplus /Zc:inline /Z7 /O2  /DNDEBUG -std:c++20 -MD /UTF8 /fp:fast /Gy /W3 /wd4200 -DPA_USE_JACK=1 -DPA_USE_ASIO=1 -DPA_USE_DS=1 -DPA_USE_WMME=1 -DPA_USE_WASAPI=1 -DPA_USE_WDMKS=1 -Zc:__cplusplus -permissive- -utf-8 /YuD:/mixxx/build/x64__portable/CMakeFiles/mixxx-lib.dir/cmake_pch.hxx /FpD:/mixxx/build/x64__portable/CMakeFiles/mixxx-lib.dir/./cmake_pch.cxx.pch /FID:/mixxx/build/x64__portable/CMakeFiles/mixxx-lib.dir/cmake_pch.hxx /showIncludes /FoCMakeFiles\mixxx-lib.dir\src\engine\sync\abletonlink.cpp.obj /Fdmixxx-lib.dir\mixxx-lib.pdb /FS -c D:\mixxx\src\engine\sync\abletonlink.cpp
D:\mixxx\src\engine\sync\abletonlink.cpp(27): error C2614: 'AbletonLink': illegal member initialization: 'm_pNumLinkPeers' is not a base or member
D:\mixxx\src\engine\sync\abletonlink.cpp(47): error C2280: 'ControlObject::ControlObject(ControlObject &&)': attempting to reference a deleted function
  D:\mixxx\src\control/controlobject.h(189): note: see declaration of 'ControlObject::ControlObject'
```
I suggest you compile it yourself first, because this two person trial and error development seems not very efficient to me.

Reply:
Okay, I've investigated and my conclusion is that moving this into the lambda is infeasible because link requires the lambda to be copy-able. This would require making the member of the lambda a `std::shared_ptr<ControlObject>` which has unnecessary overhead. They also don't allow mutable which is another oversight in the link API. (not that that is required because TIL that you can still call mutable member functions of a pointer managed by a `const unique_ptr`, which seems dangerous). 

The only recommendation I would make then is moving the `m_pNumLinkPeers` above `m_pLink` in the class definition so it doesn't get destroyed while `m_pLink` still references it.